pkg/ingester/ingester_v2_backfill.go

```diff
@@ -0,0 +1,477 @@
+	}
+}
+
+// backfillAppender is an appender to ingest old data.
+// This _does not_ implement storage.Appender interface.
+// The methods of this appender should not be called concurrently.
+type backfillAppender struct {
+	backfillTSDB                 *backfillTSDB
+	newerAppender, olderAppender *backfillRangeAppender
+
+	// Keeping this methods of Ingester here to avoid the appender having a reference to the ingester.
+	createNewTSDB func(userID, dbDir string, minBlockDuration, maxBlockDuration int64, reg *prometheus.Registry) (*userTSDB, error)
+	// We are keeping the method instead of string for backfill dir because if there is a surge in backfill writes,
+	// it will create a new string for every push while we want it only to open a new TSDB.
+	backfillBlocksDir func(userID string) string
+}
+
+type backfillRangeAppender struct {
+	app        storage.Appender
+	db         *userTSDB
+	start, end int64
+}
+
+func (a *backfillRangeAppender) within(ts int64) bool {
+	return ts >= a.start && ts < a.end
+}
+
+func newBackfillRangeAppender(db *backfillTSDBWrapper) *backfillRangeAppender {
+	return &backfillRangeAppender{
+		app:   db.db.Appender(),
+		db:    db.db,
+		start: db.start,
+		end:   db.end,
+	}
+}
+
+// add requires the samples to be within the backfill age.
+// Note: This method should not be called concurrently.
+func (a *backfillAppender) add(la []client.LabelAdapter, s client.Sample) (err error) {
+	app, db, err := a.getAppender(s)
+	if err != nil {
+		return err
+	}
+	startAppend := time.Now()
+
+	cachedRef, cachedRefExists := db.refCache.Ref(startAppend, client.FromLabelAdaptersToLabels(la))
+	// If the cached reference exists, we try to use it.
+	if cachedRefExists {
+		err = app.AddFast(cachedRef, s.TimestampMs, s.Value)
+		if err == nil || errors.Cause(err) != storage.ErrNotFound {
+			return err
+		}
+	}
+
+	// Copy the label set because both TSDB and the cache may retain it.
+	copiedLabels := client.FromLabelAdaptersToLabelsWithCopy(la)
+	ref, err := app.Add(copiedLabels, s.TimestampMs, s.Value)
+	if err == nil {
+		db.refCache.SetRef(startAppend, copiedLabels, ref)
+	}
+
+	return err
+}
+
+func (a *backfillAppender) getAppender(s client.Sample) (storage.Appender, *userTSDB, error) {
+	if !a.backfillTSDB.isWithinBackfillAge(s.TimestampMs) {
+		return nil, nil, storage.ErrOutOfBounds
+	}
+
+	// Fast path to not take the lock.
+	if a.newerAppender != nil && a.newerAppender.within(s.TimestampMs) {
+		return a.newerAppender.app, a.newerAppender.db, nil
+	} else if a.olderAppender != nil && a.olderAppender.within(s.TimestampMs) {
+		return a.olderAppender.app, a.olderAppender.db, nil
+	}
+
+	// We could use RLock here to check the TSDBs, but because we create new TSDB if none exists,
+	// it's easier and cleaner to handle races by taking the write lock.
+	// TODO(codesome): IMPORTANT: fix the issue where the compaction moves the TSDB around between 2 appends to avoid using wrong appenders in older and newer spots.
```

Comment:
Regarding this `// TODO(codesome): IMPORTANT: fix the issue where the compaction moves the TSDB around between 2 appends to avoid using wrong appenders in older and newer spots.`, I think _this is fine_ with a check for the appender being `nil` to not overwrite it and lose data.

Reasons:
1. If the older TSDB moved (i.e. outside of backfill age), that means the sample is not eligible anyway for ingestion as it went out of backfill age. So it is fine to fail the append.
2. If the newer TSDB moved to the older spot, the new appender will be for the same TSDB as before and wont be an issue ingesting to it if the sample falls in that TSDB.
3. If the newer TSDB moved to the older spot and the new sample incoming belongs to the newer TSDB to be created, then we fail that append here which will be retried later.

So, no data loss because of the re-try of Prometheus remote-write. This is a tradeoff for a whole lot complexity it would bring to handle moving appenders with TSDBs. If it turns out to be a problem, we can iterate on it later.

Reply:
I will update the comment (with `TODO`) explaining the case and that it is fine.